drivers/shared/src/make_dmy.c

```diff
@@ -66,193 +61,49 @@ make_dmy(global_param_struct *global)
                     "month (ENDMONTH), and day (ENDDAY) of the last full "
                     "simulation day");
         }
-        endday = global->endday;
-        endmonth = global->endmonth;
-        endyear = global->endyear;
-        if (LEAPYR(endyear)) {
-            days[1] = 29;
-        }
-        else {
-            days[1] = 28;
-        }
-        if (endday < days[global->endmonth - 1]) {
-            endday++;
-        }
-        else {
-            endday = 1;
-            endmonth++;
-            if (endmonth > MONTHS_PER_YEAR) {
-                endmonth = 1;
-                endyear++;
-            }
-        }
-
-        DONE = false;
-        ii = 0;
-
-        tmpyear = year;
-        tmpmonth = month;
-        tmpday = day;
-        tmphr = hr;
-        while (!DONE) {
-            get_next_time_step(&tmpyear, &tmpmonth, &tmpday, &tmphr,
-                               &tmpjday, global->dt);
-            ii++;
-            if (tmpyear == endyear) {
-                if (tmpmonth == endmonth) {
-                    if (tmpday == endday) {
-                        DONE = true;
-                    }
-                }
-            }
-        }
-        global->nrecs = ii;
+        end_dmy.day = global->endday;
+        end_dmy.month = global->endmonth;
+        end_dmy.year = global->endyear;
+        end_dmy.dayseconds = SEC_PER_DAY - global->dt;
+
+        end_num = date2num(global->time_origin_num, &end_dmy, 0.,
+                           global->calendar, TIME_UNITS_DAYS);
+        global->nrecs =
+            (unsigned)((end_num - start_num) * global->model_steps_per_day) + 1;
     }
     else {
         offset = 0;
-        tmphr = hr;
-        while (tmphr != 0) {
-            tmphr += global->dt;
+        tempsec = start_dmy.dayseconds;
+        while (tempsec != 0) {
+            tempsec += global->dt;
             offset++;
-            if (tmphr >= HOURS_PER_DAY) {
-                tmphr = 0;
+            if (tempsec >= SEC_PER_DAY) {
+                tempsec = 0;
             }
         }
```

Comment:
awkward construct. I can see you are just replacing what was before, but why not simply 

```C
for (tempsec = start_dmy.dayseconds; tempsec < SEC_PER_DAY; tempsec += global->dt) {
    offset++;
}